src/quarot/hadamard_utils.py

```diff
@@ -0,0 +1,168 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import math
+
+import torch
+from fast_hadamard_transform import hadamard_transform
+
+from .hadamard_tensors import (
+    get_had12,
+    get_had20,
+    get_had28,
+    get_had36,
+    get_had40,
+    get_had52,
+    get_had60,
+    get_had108,
+    get_had140,
+    get_had156,
+    get_had172,
+)
+
+
+def get_hadK(n, transpose=False):
+    hadK, K = None, None
+    if n % 172 == 0:  # llama-2-7b up
+        assert is_pow2(n // 172)
+        K = 172
+        hadK = get_had172().T if transpose else get_had172()
+    elif n % 156 == 0:  # llama-1-30b 3x hidden
+        assert is_pow2(n // 156)
+        K = 156
+        hadK = get_had156().T if transpose else get_had156()
+    elif n % 140 == 0:  # llama-1-30b intermediate
+        assert is_pow2(n // 140)
+        K = 140
+        hadK = get_had140().T if transpose else get_had140()
+    elif n % 108 == 0:  # llama-1-13b intermediate
+        assert is_pow2(n // 108)
+        K = 108
+        hadK = get_had108().T if transpose else get_had108()
+    elif n % 60 == 0:  # llama-1-13b 3x hidden
+        assert is_pow2(n // 60)
+        K = 60
+        hadK = get_had60().T if transpose else get_had60()
+    elif n % 52 == 0:  # llama-1-13b 1x hidden
+        assert is_pow2(n // 52)
+        K = 52
+        hadK = get_had52().T if transpose else get_had52()
+    elif n % 36 == 0:
+        assert is_pow2(n // 36)
+        K = 36
+        hadK = get_had36().T if transpose else get_had36()
+    elif n % 28 == 0:
+        assert is_pow2(n // 28)
+        K = 28
+        hadK = get_had28().T if transpose else get_had28()
+    elif n % 40 == 0:
+        assert is_pow2(n // 40)
+        K = 40
+        hadK = get_had40().T if transpose else get_had40()
+    elif n % 20 == 0:
+        assert is_pow2(n // 20)
+        K = 20
+        hadK = get_had20().T if transpose else get_had20()
```

Comment:
can we just have a single transpose at the end of the `elif`s?

```
...
elif n % 20 ==0:
    assert is_pow2(n // 12)
    K = 20
    hadK = get_had20()
...
if transpose:
    hadK = hadK.T
```

Reply:
refactored